junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TestInstancePostProcessorAndPreDestroyCallbackTests.java

```diff
@@ -0,0 +1,240 @@
+		assertBeforeAllAndAfterAllCallbacks(testClass, 1, expectedCalls);
+	}
+
+	private void assertBeforeAllAndAfterAllCallbacks(Class<?> testClass, int testsSuccessful, String... expectedCalls) {
+
+		callSequence.clear();
+
+		executeTestsForClass(testClass).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(testsSuccessful));
+
+		assertEquals(asList(expectedCalls), callSequence, () -> "wrong call sequence for " + testClass.getName());
+	}
+
+	// -------------------------------------------------------------------------
+
+	// Must NOT be private; otherwise, the @Test method gets discovered but never executed.
+	@ExtendWith({ FooClassLevelCallbacks.class, BarClassLevelCallbacks.class })
+	static class TopLevelTestCase {
+
+		@Test
+		void test() {
+			callSequence.add("test-1");
+		}
+	}
+
+	// Must NOT be private; otherwise, the @Test method gets discovered but never executed.
+	@ExtendWith(BazClassLevelCallbacks.class)
+	static class SecondLevelTestCase extends TopLevelTestCase {
+
+		@Test
+		@Override
+		void test() {
+			callSequence.add("test-2");
+		}
+	}
+
+	@ExtendWith(QuuxClassLevelCallbacks.class)
+	static class ThirdLevelTestCase extends SecondLevelTestCase {
+
+		@Test
+		@Override
+		void test() {
+			callSequence.add("test-3");
+		}
+	}
+
+	@ExtendWith(ExceptionThrowingTestInstancePreDestroyCallback.class)
+	static class ExceptionInTestInstancePreDestroyCallbackTestCase {
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	@ExtendWith(ExceptionThrowingTestInstancePostProcessor.class)
+	static class ExceptionInTestInstancePostProcessorTestCase {
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	// -------------------------------------------------------------------------
+
+	static class FooClassLevelCallbacks extends AbstractClassLevelCallbacks {
+
+		protected FooClassLevelCallbacks() {
+			super("foo");
+		}
+	}
+
+	static class BarClassLevelCallbacks extends AbstractClassLevelCallbacks {
+
+		protected BarClassLevelCallbacks() {
+			super("bar");
+		}
+	}
+
+	static class BazClassLevelCallbacks extends AbstractClassLevelCallbacks {
+
+		protected BazClassLevelCallbacks() {
+			super("baz");
+		}
+	}
+
+	static class QuuxClassLevelCallbacks extends AbstractClassLevelCallbacks {
+
+		protected QuuxClassLevelCallbacks() {
+			super("quux");
+		}
+	}
+
+	static class ExceptionThrowingTestInstancePreDestroyCallback extends AbstractClassLevelCallbacks {
+
+		protected ExceptionThrowingTestInstancePreDestroyCallback() {
+			super("exception");
+		}
+
+		@Override
+		public void preDestroyTestInstance(Object testInstance, ExtensionContext context) {
+			callSequence.add("exceptionThrowingTestInstancePreDestroyCallback");
+			throw new EnigmaException("preDestroyTestInstance");
+		}
+	}
+
+	static class ExceptionThrowingTestInstancePostProcessor extends AbstractClassLevelCallbacks {
+
+		protected ExceptionThrowingTestInstancePostProcessor() {
+			super("exception");
```

Comment:
I wouldn't put `exception` as a name here, because at line 86 it feels like the callback is actually throwing an exception when it's not.

Reply:
good point, given it an arbitrary 'foo'